src/reservoir.cxx

```diff
@@ -0,0 +1,220 @@
+
+#include "../include/reservoir.hxx"
+#include <bout/coordinates.hxx>
+#include <bout/mesh.hxx>
+#include <bout/constants.hxx>
+#include "../include/hermes_utils.hxx"  // For indexAt
+#include <iostream>   // For outputting to log file
+
+using bout::globals::mesh;
+
+// This is a simple reservoir model. It's meant for 1D but will work in 2D and 3D as well.
+// Define the reservoir location by using an analytical expression in the input file
+// e.g. reservoir_location = H(y - y_xpt) where y_xpt is the x-point location would create
+// one between the target and the X-point. 
+// The reservoir is set to a constant density and your chosen species will equilibrate with it
+// based on the density difference and a provided timescale. 
+// The intended use is to provide some form of cross-field transport for neutrals in 1D.
+
+Reservoir::Reservoir(std::string name, Options& alloptions, Solver*) 
+    :name(name){
+  AUTO_TRACE();
+
+  const Options& units = alloptions["units"];
+  const BoutReal Tnorm = units["eV"];
+  const BoutReal Nnorm = units["inv_meters_cubed"];
+  const BoutReal Omega_ci = 1. / units["seconds"].as<BoutReal>();
+
+  
+
+  // Get the options for this species
+  Options& options = alloptions[name];
+
+  reservoir_density = options["reservoir_density"]
+                  .doc("Set the density of the reservoir in [m^-3]. Default 1e19")
+                  .withDefault<BoutReal>(1e19)
+                  /Nnorm;    
+
+  reservoir_location = options["reservoir_location"]
+                  .doc("Indicates reservoir location if >0")
+                  .withDefault<Field3D>(0.0);  
+
+  reservoir_timescale = options["reservoir_timescale"]
+                  .doc("Set the timescale of the reservoir in [s]. Default 1e-6")
+                  .withDefault<BoutReal>(1e-6)
+                  *Omega_ci;
+
+  diagnose =
+      options["diagnose"].doc("Save additional diagnostics?").withDefault<bool>(false);
+          
+}
+
+
+void Reservoir::transform(Options& state) {
+  AUTO_TRACE();
+
+  // We are operating on only one species
+  auto species = state["species"][name];
```

Comment:
```suggestion
  auto& species = state["species"][name];
```

Reply:
@ZedThree question: Is there a way that we can force copies to be explicit? Perhaps delete the copy constructor and add a `copy` function so that 
```
auto species = state["species"][name];
```
would be a compile error. If someone really wants to copy they have to make it explicit:
```
auto species = copy(state["species"][name]);
```
Having the behavior depend on a single `&` in the right place is a footgun that keeps going off.